TumbleBitSetup/PoupardStern.cs

```diff
@@ -0,0 +1,263 @@
+            var Modulus = pubKey._pubKey.Modulus;
+            var Exponent = pubKey._pubKey.Exponent;
+
+            // Checking that:
+            // if y < 2^{ |N| - 1 }
+            if (!(y.CompareTo(lowerLimit) < 0))  
+                return false;
+            // if y >= 0
+            if (!(y.CompareTo(BigInteger.Zero) >= 0))
+                return false;
+            // if N > 2^{KeySize-1}
+            if (!(Modulus.CompareTo(lowerLimit) > 0))
+                return false;
+
+            // Computing K
+            GetK(k, out int BigK);
+
+            // Check if the number of x_values is not equal to K
+            if (!(xValues.Length == BigK))
+                return false;
+
+            // Get w
+            GetW(pubKey, psBytes, xValues, k, keyLength, out BigInteger w);
+
+            // Computing rPrime
+            rPrime = y.Subtract(Modulus.Multiply(w)); // Not clear if we should multiply N with W
+
+            // Encrypting and verifying the signatures
+            for (int i = 0; i < BigK; i++)
+            {
+                var z_i = SampleFromZnStar(pubKey, psBytes, i, k, keyLength);
+                // Compute right side of the equality
+                var rs = z_i.ModPow(rPrime, Modulus);
+                // If the two sides are not equal
+                if (!(xValues[i].Equals(rs)))
+                    return false;
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Generates a z_i value as specified in page 14 (3.3.1) of the setup
+        /// </summary>
+        /// <param name="pubKey">Public key used</param>
+        /// <param name="ps">public string specified in the setup</param>
+        /// <param name="i">index i</param>
+        /// <param name="k">Security parameter specified in the setup.</param>
+        /// <param name="keyLength">The size of the RSA key in bits</param>
+        /// <returns></returns>
+        internal static BigInteger SampleFromZnStar(RsaPubKey pubKey, byte[] psBytes, int i, int k, int keyLength)
+        {
+            BigInteger Modulus = pubKey._pubKey.Modulus;
+            GetK(k, out int BigK); // I think it would be more efficient to calculate K once outside this function.
+            int j = 2;
+            // Octet Length of i
+            int iLen = Utils.GetByteLength(BigK);
+            // Byte representation of i
+            var EI = Utils.I2OSP(i, iLen);
+            // ASN.1 encoding of the PublicKey
+            var keyBytes = pubKey.ToBytes();
+            // Combine the octet string
+            var combined = Utils.Combine(keyBytes, Utils.Combine(psBytes,EI));
+            for (;;)
+            {
+                // OctetLength of j
+                var jLen = Utils.GetOctetLen(j);
+                // Byte representation of j
+                var EJ = Utils.I2OSP(j, jLen);
+                // Combine EJ with the rest of the string
+                combined = Utils.Combine(combined, EJ);
```

Comment:
That's a really good catch, yes indeed it was constantly appending EJ to combined. I pushed a fix to it, please let me know what do you think now. 
4c7e12b2e53b3d82bfcc8f1329a65261c25de7a3
For the "Combine" function, it doesn't seem that there's a standard way for that. However, I just changed the combine function I was using to the one that NTumbleBit uses ([here](https://github.com/NTumbleBit/NTumbleBit/blob/master/NTumbleBit/Utils.cs#L61)) to avoid any ambiguity.